### DISCUSSION TITLE
Issue with Bitcraze Loco Positioning System - Anchor ID 7 not detected

### DISCUSSION THREAD

**Discussion by amitx64** [bitcraze/discussions#1111] [Q&A]
Hello Bitcraze community,

I recently purchased the Bitcraze Loco Explorer bundle and followed the setup steps outlined in the official documentation [here](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/). However, I encountered an issue with the Loco Positioning System, specifically with Anchor ID 7 not being detected by the Crazyflie client.

**Steps I've Taken:**

1. I did not update the Crazyflie firmware as it was preloaded. However, I tried to find the firmware for Crazyflie 2.1 on your GitHub repository but was unable to locate the release.
2. I uploaded the LPS firmware and assigned the Anchor IDs with TWR mode.
Update: While configuring the Anchor IDs, I ensured that TWR (Two-Way Ranging) mode was selected. However, upon checking the Loco Positioning Tab in the CF client (View -> Tabs -> Loco Positioning Tab), I observed that it detected TDOA (Time Difference of Arrival) instead of TWR. This is unexpected, and I would like to configure the system according to the Loco Positioning documentation, which recommends TWR mode.

**Issue:**

Upon checking the Loco Positioning Tab in the CF client (View -> Tabs -> Loco Positioning Tab), I noticed that only 7 anchors (ID 0 to ID 6) were detected and displayed with a green status. However, Anchor with ID 7 was not showing up.

**Request for Assistance:**

I'm reaching out to the community for assistance in troubleshooting this issue. I followed the steps provided by Bitcraze, but it seems that Anchor ID 7 is not being recognized. I would appreciate any insights, suggestions, or guidance on what might be causing this discrepancy.

**Additional Information:**

- I have not updated the Crazyflie firmware, as it was preloaded. However, I'm unable to find the release of the Crazyflie 2.1 firmware on your GitHub repository.
- All 8 Anchors are physically placed in the room.
- The LPS firmware has been successfully uploaded.
- Initially assigned Anchor IDs with TWR mode, but CF client is detecting TDOA.

I appreciate any help or guidance the community can provide to resolve this issue.

Thank you!

**Comment by amitx64** [comment]
@ataffanel and @knmcguire , I was able to resolve the issue by updating the firmware of the `Loco Node` and resetting the Anchor ID with Mode (TWR). Subsequently, I attempted to run the `initial_position.py` example to follow the sequence. However, after executing the script, my drone behaved erratically and eventually crashed. I also attempted to hover the Crazyflie using the cfclient takeoff button while all anchors were in place, but the drone exhibited sudden, rapid flight and collided with a wall.


I seek your assistance in identifying what might be causing these issues. Could you please advise on what I might be doing wrong and suggest corrective actions to enable proper autonomous positioning through trajectory?

**Reply by knmcguire** [reply]
Hi! Since the LPS system is not able to provide a yaw estimate but get this only from the IMU, you do have to make sure that the Crazyflie is facing the positive x direction. This is indicated in the part2 of the tutorial: https://www.bitcraze.io/documentation/tutorials/getting-started-with-flying-using-lps/

**Reply by amitx64** [reply]
@knmcguire & @ataffanel: Hi, I just wanted to inform you that, after following all the instructions provided by `Bitcraze` regarding the `Loco position system`, I aimed to observe some `positioning` behaviors such as `hovering` and `sequencing`. To achieve this, I discovered that Bitcraze has shared some examples in the `crazyflie-lib-python` repository, one of which is the `initial_position.py` script.
```
import math
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.crazyflie.syncLogger import SyncLogger
from cflib.utils import uri_helper

# URI to the Crazyflie to connect to
uri = uri_helper.uri_from_env(default='radio://0/80/2M/E7E7E7E7E7')

# Change the sequence according to your setup
#             x    y    z
sequence = [
    (0, 0, 0.7),
    (-0.7, 0, 0.7),
    (0, 0, 0.7),
    (0, 0, 0.2),
]


def wait_for_position_estimator(scf):
    print('Waiting for estimator to find position...')

    log_config = LogConfig(name='Kalman Variance', period_in_ms=500)
    log_config.add_variable('kalman.varPX', 'float')
    log_config.add_variable('kalman.varPY', 'float')
    log_config.add_variable('kalman.varPZ', 'float')

    var_y_history = [1000] * 10
    var_x_history = [1000] * 10
    var_z_history = [1000] * 10

    threshold = 0.001

    with SyncLogger(scf, log_config) as logger:
        for log_entry in logger:
            data = log_entry[1]

            var_x_history.append(data['kalman.varPX'])
            var_x_history.pop(0)
            var_y_history.append(data['kalman.varPY'])
            var_y_history.pop(0)
            var_z_history.append(data['kalman.varPZ'])
            var_z_history.pop(0)

            min_x = min(var_x_history)
            max_x = max(var_x_history)
            min_y = min(var_y_history)
            max_y = max(var_y_history)
            min_z = min(var_z_history)
            max_z = max(var_z_history)

            # print("{} {} {}".
            #       format(max_x - min_x, max_y - min_y, max_z - min_z))

            if (max_x - min_x) < threshold and (
                    max_y - min_y) < threshold and (
                    max_z - min_z) < threshold:
                break


def set_initial_position(scf, x, y, z, yaw_deg):
    scf.cf.param.set_value('kalman.initialX', x)
    scf.cf.param.set_value('kalman.initialY', y)
    scf.cf.param.set_value('kalman.initialZ', z)

    yaw_radians = math.radians(yaw_deg)
    scf.cf.param.set_value('kalman.initialYaw', yaw_radians)


def reset_estimator(scf):
    cf = scf.cf
    cf.param.set_value('kalman.resetEstimation', '1')
    time.sleep(0.1)
    cf.param.set_value('kalman.resetEstimation', '0')

    wait_for_position_estimator(cf)


def run_sequence(scf, sequence, base_x, base_y, base_z, yaw):
    cf = scf.cf

    for position in sequence:
        print('Setting position {}'.format(position))

        x = position[0] + base_x
        y = position[1] + base_y
        z = position[2] + base_z

        for i in range(50):
            cf.commander.send_position_setpoint(x, y, z, yaw)
            time.sleep(0.1)

    cf.commander.send_stop_setpoint()
    # Hand control over to the high level commander to avoid timeout and locking of the Crazyflie
    cf.commander.send_notify_setpoint_stop()

    # Make sure that the last packet leaves before the link is closed
    # since the message queue is not flushed before closing
    time.sleep(0.1)


if __name__ == '__main__':
    cflib.crtp.init_drivers()

    # Set these to the position and yaw based on how your Crazyflie is placed
    # on the floor
    initial_x = 1.0
    initial_y = 1.0
    initial_z = 0.0
    initial_yaw = 90  # In degrees
    # 0: positive X direction
    # 90: positive Y direction
    # 180: negative X direction
    # 270: negative Y direction

    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        set_initial_position(scf, initial_x, initial_y, initial_z, initial_yaw)
        reset_estimator(scf)
        run_sequence(scf, sequence,
                     initial_x, initial_y, initial_z, initial_yaw)
```
However, after executing the script, my drone behaved erratically and eventually crashed. Subsequently, in order to achieve stable hovering with the Crazyflie, I made some modifications to the script. Here is the updated version:
```
# Changed sequence according to your setup
#             x    y    z
sequence = [
    (0, 0, 0.7),
]

if __name__ == '__main__':
    cflib.crtp.init_drivers()

    # Set these to the position and yaw based on how your Crazyflie is placed
    # on the floor
    initial_x = 0.0
    initial_y = 0.0
    initial_z = 0.2
    initial_yaw = 90  # In degrees
    # 0: positive X direction
    # 90: positive Y direction
    # 180: negative X direction
```
Unfortunately, when running this modified script, my drone reached the desired altitude but did not hover steadily; instead, it exhibited continuous drifting in one direction. I thought the `Crazyflie` would correct its position error to achieve the target position.
   
`Any insights or suggestions on how to address this issue would be greatly appreciated.`

**Reply by knmcguire** [reply]
Hmm interesting. Unfortunately I wouldn't know what is wrong with the script as it looks fine to me. Something that comes to mind is that perhaps the crazyflie is not facing positive 90 degrees in relation to the LPS system but negative 90 degrees (positive is turning left). You can double-check that in the coordinate system here: https://www.bitcraze.io/documentation/system/platform/cf2-coordinate-system/ but make sure to try both directions (maybe there is a minus off somewhere).

If not, then we have to find time to check if this functionality still works as it might be broken, but if so I'll make an issue at the right repo for that.

**Reply by amitx64** [reply]
Thank you for your response. I appreciate your assistance. I have configured everything according to the Bitcraze documentation for the `Loco Position Explorer bundle`. However, I am `encountering issues with positioning`. If possible, `could you provide an example script` or `guide me on how to achieve positioning using Crazyflie`? `My objective is to set a target position for the Crazyflie and have it follow my instructions to reach that position`. Your help in this matter would be greatly appreciated. Thank you.

**Reply by knmcguire** [reply]
Ah oke sorry, so you haven't achieved any flight whatsoever then? From some posts before it seemed that you had but I misunderstood it.

I don't think that initial position script is a good example for flight. You should rather have a look at the examples in /autonomy/ but it is better to test first test the flight out command-based flight on the flight tab of the client: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/flightcontrol_tab/. That one you can try out take off and hover with just a push of a button. That needs to work before anything else.

If that crashes, then first try to debug the situation by holding the crazyflie in your hand and move it around, see if that corresponds with the position estimates on the loco positioning tab

**Reply by amitx64** [reply]
@knmcguire When I connected the `Crazyflie` with the `CFClient`, I noticed that the `Command-Based Control Flight` is not active. I don't know why this is happening. Additionally, I would like to inquire if there is an `alternative to sending position commands` through the `Python script`. If possible, could you provide a simple example of `hardcoding the sequence in the Crazyflie firmware`? I believe this approach might yield better performance.

Please guide me in the right direction. 
Thank you.

**Reply by knmcguire** [reply]
That is actually concerning. The command based control flight is active whenever a positioning deck is detected and in your case it means that it is not detecting the locodeck. 

Could you share the console tab output?

**Reply by amitx64** [reply]
I am sharing the "screen shots":
1. "Flight Control":
![Screenshot 2024-01-09 161907](https://github.com/bitcraze/discussions/assets/79006087/41caa492-3ef2-424a-9714-120345e64e8e)

2. "Loco Positioning":
![Screenshot 2024-01-09 161845](https://github.com/bitcraze/discussions/assets/79006087/3f8767a7-2b0d-4735-bcbe-9ab0a3ac662e)

**Reply by amitx64** [reply]
According to my analysis, the `Crazyflie` does not perform any position correction when we pass position coordinates. I mean, the `Crazyflie` doesn't stay at the initially given position coordinates; instead, it starts drifting from its initial position. We initially assumed that it would remain at the provided position until we provided the next set of position coordinates. We are currently using the `initial_position.py` example from `Bitcraze`.

**Reply by knmcguire** [reply]
What I mean with console tab output is this one: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/console_tab/. You can find it by view->toolboxes and you can just copy past the content of that in here (but do use the code brackets)

**Reply by ataffanel** [reply]
Reading the first comment in the script, your assumption is not what this script is doing. It is for setting an initial position estimate in the Kalman filter for very specific cases where the default behaviour is not sufficient (ie. when not starting with yaw at 0). It is normal that the kalman filter continues estimating from this initial position and if your system is not stable or not well setup it is normal that the drift continues after that. It is very likely that you do not need this script at this point.

In a loco system, you do not pass position to the Crazyflie for position estimation, all measurements needed to estimate the position comes from the UWB radio measurements. So if your position estimate is not correct, you need to make sure the loco positioning system is setup correctly and that you start the Crazyflie facing X (with a yaw as close as possible to 0).

**Reply by amitx64** [reply]
@knmcguire Here is the console tab screenshot. I took the screenshot while loco-nodes were turned off due to low battery. I am also in the process of checking the voltage divider calibration in the nrf-51822 firmware, which is why it is displaying the wrong voltage for the Crazyflie.
![console_image](https://github.com/bitcraze/discussions/assets/79006087/f3e9c295-0242-4bd0-b906-fb041cade1b0)

**Reply by knmcguire** [reply]
The supervisor is indicating that the crazyflie can not fly but there is no more information. If you install the latest cfclient with `pip install cfclient --upgrade`, then you will be able to see more information of why the supervisor is blocking the flight entirely in the flight tab.

Also it would be helpful to run the latest release of both nrf and stm without any changes.

**Reply by amitx64** [reply]
@knmcguire Actually, I forgot to mention that I have also disabled auto `ARM` in the `Crazyflie`. That's why this issue is occurring. I already have the latest `CFClient` and `firmware`.

**Reply by knmcguire** [reply]
Hi, So in the console I can see that there is a modified STM and NRF firmware on top of the latest release, so if you could flash the latest release of 2023.11 that would be helpful. Also I can see that you don't have the latest cfclient, as it is missing the supervisor field in the flight tab (see this [blogpost](https://www.bitcraze.io/2023/12/new-release-2023-11/)). We need to be able to determine if the system is still working for you in a situation that is reproducible for us to test out with. This way we can determine what the issue is without the influence of any made modification on your end, or else we will not be able to help  you properly
